pkg/issuer/vault/vault.go

```diff
@@ -0,0 +1,119 @@
+	errorVaultConfigRequired        = "ErrVaultConfigRequired"
+	errorVaultServerAndPathRequired = "ErrVaultServerAndPathRequired"
+	errorVaultAuthFieldsRequired    = "ErrVaultAuthFieldsRequired"
+	errorVaultDurationInvalid       = "ErrVaultDurationInvalid"
+
+	messageNamespaceRequired     = "resource namespace cannot be empty"
+	messageVaultConfigRequired   = "Vault config cannot be empty"
+	messageServerAndPathRequired = "Vault server and path are required fields"
+	messsageAuthFieldsRequired   = "Vault tokenSecretRef or appRole is required"
+	messageAuthFieldRequired     = "Vault tokenSecretRef and appRole cannot be set on the same issuer"
+	messageVaultDurationInvalid  = "Vault %s"
+)
+
+type Vault struct {
+	issuer v1alpha1.GenericIssuer
+
+	client   kubernetes.Interface
+	cmclient clientset.Interface
+	recorder record.EventRecorder
+
+	secretsLister corelisters.SecretLister
+
+	// issuerResourcesNamespace is a namespace to store resources in. This is
+	// here so we can easily support ClusterIssuers with the same codepath. By
+	// setting this field to either the namespace of the Issuer, or the
+	// clusterResourceNamespace specified on the CLI, we can easily continue
+	// to work with supplemental (e.g. secrets) resources without significant
+	// refactoring.
+	issuerResourcesNamespace string
+}
+
+func NewVault(issuerObj v1alpha1.GenericIssuer,
+	cl kubernetes.Interface,
+	cmclient clientset.Interface,
+	recorder record.EventRecorder,
+	resourceNamespace string,
+	secretsLister corelisters.SecretLister) (issuer.Interface, error) {
+
+	if resourceNamespace == "" {
+		return nil, updateEventAndCondition(fmt.Errorf(messageNamespaceRequired), errorVaultNamespaceRequired, issuerObj, recorder)
+	}
+
+	if issuerObj.GetSpec().Vault == nil {
+		return nil, updateEventAndCondition(fmt.Errorf(messageVaultConfigRequired), errorVaultConfigRequired, issuerObj, recorder)
+	}
+
+	if issuerObj.GetSpec().Vault.Server == "" ||
+		issuerObj.GetSpec().Vault.Path == "" {
+		return nil, updateEventAndCondition(fmt.Errorf(messageServerAndPathRequired), errorVaultServerAndPathRequired, issuerObj, recorder)
+	}
+
+	if issuerObj.GetSpec().Vault.Auth.TokenSecretRef.Name == "" &&
+		issuerObj.GetSpec().Vault.Auth.AppRole.RoleId == "" &&
+		issuerObj.GetSpec().Vault.Auth.AppRole.SecretRef.Name == "" {
+		return nil, updateEventAndCondition(fmt.Errorf(messsageAuthFieldsRequired), errorVaultAuthFieldsRequired, issuerObj, recorder)
+	}
+
+	if issuerObj.GetSpec().Vault.Auth.TokenSecretRef.Name != "" &&
+		(issuerObj.GetSpec().Vault.Auth.AppRole.RoleId != "" ||
+			issuerObj.GetSpec().Vault.Auth.AppRole.SecretRef.Name != "") {
+		return nil, updateEventAndCondition(fmt.Errorf(messageAuthFieldRequired), errorVaultAuthFieldsRequired, issuerObj, recorder)
+	}
+
+	if err := issuer.ValidateDuration(issuerObj); err != nil {
+		return nil, updateEventAndCondition(fmt.Errorf(messageVaultDurationInvalid, err.Error()), errorVaultDurationInvalid, issuerObj, recorder)
+	}
```

Comment:
All validation of the Issuer resource should happen in Setup() - other fields (eg resourceNamespace) can be validated here, but not the issuer itself.